src/mlpack/methods/ann/layer/instance_norm_impl.hpp

```diff
@@ -0,0 +1,143 @@
+/**
+ * @file methods/ann/layer/instance_norm_impl.hpp
+ * @author Anjishnu Mukherjee
+ *
+ * Implementation of the Instance Normalization Layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_INSTANCENORM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_INSTANCENORM_IMPL_HPP
+
+// In case it is not included.
+#include "instance_norm.hpp"
+
+namespace mlpack {
+namespace ann { /** Artificial Neural Network. */
+
+template<typename InputDataType, typename OutputDataType>
+InstanceNorm<InputDataType, OutputDataType>::InstanceNorm() :
+    size(0),
+    eps(1e-8),
+    average(true),
+    momentum(0.0),
+    deterministic(false),
+    reset(false)
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+InstanceNorm<InputDataType, OutputDataType>::InstanceNorm(
+    const size_t size,
+    const size_t numFunctions,
+    const double eps,
+    const bool average,
+    const double momentum) :
+    size(size),
+    numFunctions(numFunctions),
+    eps(eps),
+    average(average),
+    momentum(momentum),
+    deterministic(false),
+    reset(false)
+{
+    batchNorm = ann::BatchNorm<> (size * numFunctions, eps, average, momentum);
```

Comment:
Honestly I'm not sure I follow here.  Let me know if I overlooked something.  Following the paper: https://arxiv.org/pdf/1607.08022.pdf, it seems to me that instance normalization is applied individually for each input point, as in Eq. (3) in the paper.  Batch normalization is described in Eq. (2).

Specifically, I read batch normalization's mean and variance as computed via three summations: over the batch, over the columns of the image, and over the rows of the image.  But instance normalization is computed only via two: over the columns, and over the rows.  So I am not sure why we are expressing our computation here in terms of `BatchNorm`---it seems like it would be easier to express it just as two summations.  I also don't get why `numFunctions` is necessary.

Please note that I have only quickly skimmed/half-read the paper to try and keep this PR moving, so like I said, it's completely possible I have overlooked something here.  Let me know what you think, and I'm happy to dig deeper in the paper if needed. :)

Reply:
**Regarding why this was implemented as a batch norm wrapper :**

```Notations used -> N : Batch Size, C : Number of channels, H : Height, W : Width```
Instance Norm applied on a (N,C,H,W) input is equivalent to BatchNorm applied on a (1,N*C,H,W) input. 
This idea is same as the one followed by the original author of the paper [here](https://github.com/DmitryUlyanov/texture_nets/blob/master/InstanceNormalization.lua) and also the same as the PyTorch implementation [here](https://github.com/pytorch/pytorch/pull/1283).

I guess the reason for doing this is mainly because it will allow us to reuse the batch norm backward and gradient functions. We could have written the forward method as two summations, as you mentioned (that was my initial approach for my original PR), but then we would also have to write the backward and gradient methods from scratch (?), unless I am missing something.

**Regarding the numFunctions parameter :**

Because this implementation uses the (N,C,H,W) -> (1,N*C,H,W) mapping, batchnorm needs to be instantiated like ```ann::BatchNorm<> (size*input.n_cols, eps, average, momentum)```. Previously, I was doing it in the Forward method so when ResetParameters() is called by FFN, memory isn't allotted for the parameters ( & subsequently gradient) of InstanceNorm layer. Hence, there was a crash happening during the gradient test. 
@hello-fri-end figured out this bug, and came up with the solution to instantiate the batchnorm object in the constructor itself. To do so, we need to pass ```input.n_cols``` as an argument.

Even I am not very sure about whether having this as an argument makes sense. But I guess the only way we can avoid this argument, is if we re-write the entire instance norm layer from scratch instead of writing it as a batchnorm wrapper.

Reply:
Thanks for the explanation @iamshnoo. Although, I'm not sure if implementing the layer from scratch would allow us to avoid the `numFunctions ` parameter because ultimately, `ResetParameters()` will expect `WeightSize()` to return the total size of the weights ( which in this case is `size * input.n_cols`) . So, either way, we somehow need to know the number of input columns before the `Forward()` function.

I could be wrong about some of this. Please let me know if I've overlooked something.

Reply:
Thanks for the explanation!  I still have one piece that I don't understand, though.  In batch normalization, we learn parameters `gamma` and `beta` for each dimension `(C, H, W)` of the input (here I'm just assuming the input is an image, which is fine, you could easily adapt to other cases too), and this `gamma` and `beta` are applied to the output as `y_{c,h,w} = gamma_{c,h,w} * x'_{c,h,w} + beta_{c,h,w}`, where `x'_{c,h,w}` is the batch-normalized version of the point `x_{c,h,w}`.  (I'm basing this off Algorithm 1 in the batch norm paper: http://proceedings.mlr.press/v37/ioffe15.pdf)

This implementation of instance normalization that you are proposing, though, effectively learns one `gamma` and `beta` for each input point.  So if we are using a batch normalization of shape (1, N*C, H, W) as our instance norm implementation, we end up doing this:

```
y_{n,c,h,w} = gamma_{n,c,h,w} * x'_{n,c,h,w} + beta_{n,c,h,w}
```

But that intuitively means that we learn a `gamma` and `beta` for each point in a batch... or, stated differently, we end up with a network where the same point might be transformed differently if it's the 2nd point in a batch vs. being the 1st point in a batch.  To me I don't think that's right.  So, I think maybe what we want here is `gamma` and `beta` to be learned only on the pixels of the image `{c,h,w}`, and not including the position in the batch `n`.

So, e.g., we'd want this:

```
y_{n,c,h,w} = gamma_{c,h,w} * x'_{n,c,h,w} + beta_{c,h,w}
```

Please correct me if I missed something.  If the formulation I just wrote is correct, then the size of the parameters is always just the size of `gamma` plus the size of `beta`, which should be `C*H*W`, not `N*C*H*W`---that is, the size of the weights should not depend on the batch size.

However, if that is the case, I am not sure if we can use the batch normalization implementation as-is.  (Perhaps we could modify it slightly so it works both for instance normalization and batch normalization.)  Let me know what you think. :+1:

Reply:
Thanks for looking into this @rcurtin. I'm also unsure how affine transform is supposed to be applied to the normalized points since there is no mention of it in the [original paper](https://arxiv.org/pdf/1607.08022.pdf). Although, I took a look at the Pytorch implementation [https://github.com/pytorch/pytorch/pull/1283/files#diff-d49c3b3b20727c953a6f14cba34746844c3e617ec4a9016a398919430d766491R21](https://github.com/pytorch/pytorch/pull/1283/files#diff-d49c3b3b20727c953a6f14cba34746844c3e617ec4a9016a398919430d766491R21) & their implemenation also learns a `gamma` & `beta` for each point in the batch. Since, Instance Normalization is generally used in image synthesis & style transfer, maybe the idea is to enable the model to have `N` different behaviors conditioned on a style class label ` i , i ∈ (1, . . . , N)`. Maybe, we should add a boolean parameter for affine transform in ` BatchNorm`  &  `InstanceNorm ` so that users have an option of disabling it if they want.

Reply:
Wow, fascinating.  I guess, intuitively, I don't understand why we would learn one weight for each element in a batch---to me this doesn't make sense.  What you said about `N` different behaviors makes sense in principle, but what I don't understand is how the position in the batch could be used as some kind of style class label.

Anyway, it is clear that the reference implementation does learn one weight for each element in a batch, and it's probably good to stick to the paper's description and the reference implementation.  In this case, it seems like the solution you have proposed is the best way to go.  I might suggest changing the name from `numFunctions` to `batchSize`, and then also just adding a quick assertion or check in `Forward()` that the given matrix has the correct shape, and if it doesn't, throwing some exception or something like this to let the user know that they have to use the same `batchSize` that was used in the constructor.